File: backend/app/schemas/analysis.py
"""
Analysis request and response schemas.

Mirrors and extends the VerificationResult interface from the frontend types.ts.
"""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field, field_validator

from .evidence import EvidenceItem


# ── Enums ────────────────────────────────────────────────────────────────


class Language(str, Enum):
    """Supported analysis languages."""

    AUTO = "auto"
    EN = "en"
    HI = "hi"
    TE = "te"


class Assessment(str, Enum):
    """Possible verdict outcomes."""

    SUPPORTED = "SUPPORTED"
    POTENTIALLY_MISLEADING = "POTENTIALLY_MISLEADING"
    INSUFFICIENT_EVIDENCE = "INSUFFICIENT_EVIDENCE"
    CONFLICTING_EVIDENCE = "CONFLICTING_EVIDENCE"


class EvidenceStrength(str, Enum):
    """Qualitative strength of the assembled evidence."""

    WEAK = "WEAK"
    MODERATE = "MODERATE"
    STRONG = "STRONG"


class PipelineStatus(str, Enum):
    """Overall status of the analysis pipeline."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


# ── Request ──────────────────────────────────────────────────────────────


class AnalysisRequest(BaseModel):
    """Incoming analysis request from the frontend."""

    text: str = Field(
        ...,
        min_length=1,
        description="The text to analyze — a claim, article, social-media post, or message",
    )
    language: Language = Field(
        default=Language.AUTO,
        description="Language of the input text, or 'auto' for detection",
    )

    @field_validator("text")
    @classmethod
    def text_must_not_be_blank(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("Input text must not be blank")
        return stripped

    model_config = {"json_schema_extra": {"example": {
        "text": "India has banned UPI payments and the service has been shut down.",
        "language": "auto",
    }}}


# ── Confidence Breakdown ─────────────────────────────────────────────────


class ConfidenceBreakdown(BaseModel):
    """Granular confidence sub-scores (populated when real models run)."""

    evidence_agreement: Optional[float] = Field(None, ge=0.0, le=1.0)
    evidence_relevance: Optional[float] = Field(None, ge=0.0, le=1.0)
    source_quality: Optional[float] = Field(None, ge=0.0, le=1.0)
    model_confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    probabilities: Optional[dict[str, float]] = Field(None, description="Per-class probability distribution")



# ── Response ─────────────────────────────────────────────────────────────


class AnalysisResponse(BaseModel):
    """Full analysis result returned to the frontend."""

    analysis_id: str = Field(
        ..., description="Unique identifier for this analysis run"
    )
    language: str = Field(
        ..., description="Detected or specified language code"
    )
    claim: str = Field(
        ..., description="The factual claim extracted from the input"
    )
    claim_type: Optional[str] = Field(
        None, description="Classification of the claim (e.g. 'Statistical claim')"
    )
    assessment: Assessment = Field(
        ..., description="Verdict of the analysis"
    )
    confidence: Optional[float] = Field(
        None, ge=0.0, le=1.0, description="Overall confidence score (null until real models run)"
    )
    confidence_breakdown: Optional[ConfidenceBreakdown] = Field(
        None, description="Granular confidence sub-scores"
    )
    confidence_tier: Optional[str] = Field(
        None, description="Qualitative confidence label (HIGH CONFIDENCE, MODERATE CONFIDENCE, LOW CONFIDENCE)"
    )
    confidence_explanation: Optional[str] = Field(
        None, description="Transparent explanation of uncertainty and signal fusion"
    )
    evidence_strength: Optional[EvidenceStrength] = Field(
        None, description="Qualitative strength of assembled evidence"
    )
    evidence: list[EvidenceItem] = Field(
        default_factory=list, description="Evidence items traced for this claim"
    )
    explanation: str = Field(
        ..., description="Human-readable explanation of the assessment"
    )
    pipeline_status: PipelineStatus = Field(
        ..., description="Status of the analysis pipeline"
    )

    model_config = {"json_schema_extra": {"example": {
        "analysis_id": "a1b2c3d4e5f6",
        "language": "en",
        "claim": "India has banned UPI payments.",
        "claim_type": None,
        "assessment": "INSUFFICIENT_EVIDENCE",
        "confidence": None,
        "confidence_breakdown": None,
        "evidence_strength": None,
        "evidence": [],
        "explanation": "No AI model or evidence retrieval service is configured yet. "
                       "This is a placeholder response from the analysis pipeline.",
        "pipeline_status": "completed",
    }}}


# ── Summary ──────────────────────────────────────────────────────────────


class AnalysisSummary(BaseModel):
    """Compact summary of an analysis for listing in history."""

    analysis_id: str = Field(..., description="Unique identifier for this analysis")
    language: str = Field(..., description="Language code")
    claim: str = Field(..., description="The primary claim text")
    claim_type: Optional[str] = Field(None, description="Classification of the claim")
    assessment: Assessment = Field(..., description="Verdict outcome")
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0, description="Model confidence score")
    confidence_tier: Optional[str] = Field(None, description="Confidence tier label")
    evidence_strength: Optional[EvidenceStrength] = Field(None, description="Evidence strength")
    evidence_count: int = Field(0, description="Total evidence items traced")
    created_at: Optional[str] = Field(None, description="ISO timestamp of when the analysis was performed")


# ── Error ────────────────────────────────────────────────────────────────


class ErrorResponse(BaseModel):
    """Structured error returned on validation or processing failure."""

    error: str = Field(..., description="Machine-readable error code")
    message: str = Field(..., description="Human-readable error message")
    request_id: Optional[str] = Field(None, description="Request trace ID")

File: src/services/api/analysis.ts
/**
 * VeriTrace Analysis API Service
 *
 * Implements communication with /api/v1/analyze and /api/v1/analysis/{id}.
 * Maps backend responses cleanly to the frontend VerificationResult model.
 */

import type { Language, VerificationResult, EvidenceItem, Relation } from "../../types";
import { apiFetch } from "./client";

export interface BackendEvidenceItem {
  id: string;
  title: string;
  source: string;
  publisher?: string | null;
  url: string;
  snippet: string;
  relevance_score: number;
  source_quality: number;
  relation: Relation;
  published_at?: string | null;
}

export interface BackendConfidenceBreakdown {
  evidence_agreement?: number | null;
  evidence_relevance?: number | null;
  source_quality?: number | null;
  model_confidence?: number | null;
  probabilities?: Record<string, number> | null;
}

export interface BackendAnalysisResponse {
  analysis_id: string;
  language: string;
  claim: string;
  claim_type?: string | null;
  assessment: VerificationResult["assessment"];
  confidence?: number | null;
  confidence_tier?: string | null;
  confidence_explanation?: string | null;
  confidence_breakdown?: BackendConfidenceBreakdown | null;
  evidence_strength?: VerificationResult["evidenceStrength"] | null;
  evidence: BackendEvidenceItem[];
  explanation: string;
  pipeline_status: "pending" | "running" | "completed" | "failed";
}

export interface AnalyzeOptions {
  language?: "auto" | "en" | "hi" | "te" | string;
  signal?: AbortSignal;
  timeoutMs?: number;
}

/**
 * Map raw backend analysis response to frontend VerificationResult.
 */
export function mapBackendResponseToVerificationResult(
  data: BackendAnalysisResponse,
): VerificationResult {
  const lang = (["en", "hi", "te"].includes(data.language)
    ? data.language
    : "en") as Language;

  const mappedEvidence: EvidenceItem[] = (data.evidence || []).map((ev) => ({
    id: ev.id,
    title: ev.title,
    source: ev.publisher || ev.source || "Unknown Source",
    publisher: ev.publisher || ev.source || undefined,
    url: ev.url,
    snippet: ev.snippet,
    relevanceScore: ev.relevance_score,
    sourceQuality: ev.source_quality,
    relation: ev.relation,
    publishedAt: ev.published_at || undefined,
  }));

  const bd = data.confidence_breakdown;
  const confidenceBreakdown = bd
    ? {
        evidenceAgreement: bd.evidence_agreement ?? undefined,
        evidenceRelevance: bd.evidence_relevance ?? undefined,
        sourceQuality: bd.source_quality ?? undefined,
        modelConfidence: bd.model_confidence ?? undefined,
        probabilities: bd.probabilities ?? undefined,
      }
    : undefined;

  return {
    analysisId: data.analysis_id,
    claim: data.claim,
    language: lang,
    claimType: data.claim_type || "Factual claim",
    assessment: data.assessment,
    confidence: data.confidence !== undefined ? data.confidence : null,
    confidenceTier: data.confidence_tier || undefined,
    confidenceExplanation: data.confidence_explanation || undefined,
    confidenceBreakdown,
    evidenceStrength: data.evidence_strength || undefined,
    explanation: data.explanation,
    evidence: mappedEvidence,
  };
}

/**
 * Submit text to POST /api/v1/analyze on the real backend.
 */
export async function analyzeClaim(
  text: string,
  options: AnalyzeOptions = {},
): Promise<VerificationResult> {
  const trimmed = text.trim();
  if (!trimmed) {
    throw new Error("Enter a claim, article or message to verify.");
  }

  const payload = {
    text: trimmed,
    language: options.language || "auto",
  };

  const response = await apiFetch<BackendAnalysisResponse>("/api/v1/analyze", {
    method: "POST",
    body: payload,
    signal: options.signal,
    timeoutMs: options.timeoutMs ?? 15000,
  });

  return mapBackendResponseToVerificationResult(response);
}

/**
 * Retrieve a previously completed analysis by ID from GET /api/v1/analysis/{id}.
 */
export async function getAnalysisById(
  analysisId: string,
  options: { signal?: AbortSignal; timeoutMs?: number } = {},
): Promise<VerificationResult> {
  const response = await apiFetch<BackendAnalysisResponse>(
    `/api/v1/analysis/${encodeURIComponent(analysisId)}`,
    {
      method: "GET",
      signal: options.signal,
      timeoutMs: options.timeoutMs ?? 10000,
    },
  );

  return mapBackendResponseToVerificationResult(response);
}

export interface BackendAnalysisSummary {
  analysis_id: string;
  language: string;
  claim: string;
  claim_type?: string | null;
  assessment: VerificationResult["assessment"];
  confidence?: number | null;
  confidence_tier?: string | null;
  evidence_strength?: VerificationResult["evidenceStrength"] | null;
  evidence_count: number;
  created_at?: string | null;
}

/**
 * Fetch a list of previously completed analyses from GET /api/v1/analyses.
 */
export async function listAnalyses(
  options: { limit?: number; offset?: number; signal?: AbortSignal; timeoutMs?: number } = {},
): Promise<import("../../types").AnalysisSummary[]> {
  const params = new URLSearchParams();
  if (options.limit) params.set("limit", String(options.limit));
  if (options.offset) params.set("offset", String(options.offset));
  const query = params.toString() ? `?${params.toString()}` : "";

  const response = await apiFetch<BackendAnalysisSummary[]>(`/api/v1/analyses${query}`, {
    method: "GET",
    signal: options.signal,
    timeoutMs: options.timeoutMs ?? 10000,
  });

  return response.map((item) => ({
    analysisId: item.analysis_id,
    language: (["en", "hi", "te"].includes(item.language) ? item.language : "en") as Language,
    claim: item.claim,
    claimType: item.claim_type || "Factual claim",
    assessment: item.assessment,
    confidence: item.confidence ?? null,
    confidenceTier: item.confidence_tier || undefined,
    evidenceStrength: item.evidence_strength || undefined,
    evidenceCount: item.evidence_count,
    createdAt: item.created_at || undefined,
  }));
}
